samples/bluetooth/peripheral_uart/src/uart_async_adapter.c

```diff
@@ -0,0 +1,551 @@
+ *
+ * @param dev UART device structure.
+ * @param evt Pointer to uart_event structure.
+ */
+static void user_callback(const struct device *dev, struct uart_event *evt)
+{
+	struct uart_async_adapter_data *data = access_dev_data(dev);
+
+	if (data->user_callback) {
+		data->user_callback(dev, evt, data->user_data);
+	}
+}
+
+
+
+static int callback_set(const struct device *dev, uart_callback_t callback, void *user_data)
+{
+	struct uart_async_adapter_data *data = access_dev_data(dev);
+
+	data->user_callback = callback;
+	data->user_data = user_data;
+
+	return 0;
+}
+
+static inline void notify_rx_buffer(const struct device *dev)
+{
+	struct uart_event event = {UART_RX_RDY};
+	struct uart_async_adapter_data *data = access_dev_data(dev);
+	bool send = false;
+	k_spinlock_key_t key = k_spin_lock(&(data->lock));
+
+	if (data->rx.buf) {
+
+		__ASSERT_NO_MSG(data->rx.curr_buf >= data->rx.buf);
+		__ASSERT_NO_MSG(data->rx.curr_buf >= data->rx.last_notify_buf);
+		if (data->rx.curr_buf > data->rx.last_notify_buf) {
+
+			event.data.rx.buf = data->rx.buf;
+			event.data.rx.len = data->rx.curr_buf - data->rx.last_notify_buf;
+			event.data.rx.offset = data->rx.last_notify_buf - data->rx.buf;
+			data->rx.last_notify_buf = data->rx.curr_buf;
+			send = true;
+		}
+	}
+	k_spin_unlock(&(data->lock), key);
+	if (send) {
+		LOG_DBG("Notification: UART_RX_RDY (0x%x + %d, size: %d)",
+			(unsigned int)event.data.rx.buf, event.data.rx.offset, event.data.rx.len);
+		user_callback(dev, &event);
+	}
+}
+
+static inline void notify_rx_buffer_release(const struct device *dev, uint8_t *buf)
+{
+	struct uart_event event = {
+		.type = UART_RX_BUF_RELEASED,
+		.data.rx_buf.buf = buf
+	};
+	LOG_DBG("Notification: UART_RX_BUF_RELEASED");
+	user_callback(dev, &event);
+}
+
+static inline void notify_rx_buffer_req(const struct device *dev)
+{
+	struct uart_event event = {
+		.type = UART_RX_BUF_REQUEST,
+	};
+	LOG_DBG("Notification: UART_RX_BUF_REQUEST");
+	user_callback(dev, &event);
+}
+
+static inline void switch_rx_buffer(const struct device *dev, bool request_new)
+{
+	struct uart_async_adapter_data *data = access_dev_data(dev);
+	k_spinlock_key_t key = k_spin_lock(&(data->lock));
+	uint8_t *old_buf = old_buf = data->rx.buf;
+
+	data->rx.buf = data->rx.next_buf;
+	data->rx.curr_buf = data->rx.next_buf;
+	data->rx.last_notify_buf = data->rx.next_buf;
+	data->rx.size_left = data->rx.next_buf_len;
+	data->rx.next_buf = NULL;
+	data->rx.next_buf_len = 0;
+	k_spin_unlock(&(data->lock), key);
+
+	if (old_buf) {
+		notify_rx_buffer_release(dev, old_buf);
+	}
+	if (request_new) {
+		notify_rx_buffer_req(dev);
+	}
+}
+
+static int tx(const struct device *dev, const uint8_t *buf, size_t len, int32_t timeout)
+{
+	int ret = 0;
+	struct uart_async_adapter_data *data = access_dev_data(dev);
+	bool tx_send = false;
+
+	LOG_DBG("%s(%s, 0x%x, %u, %d)", __func__, dev->name, (unsigned int)buf, len, timeout);
+
+	k_spinlock_key_t key = k_spin_lock(&(data->lock));
+
```

Comment:
Separation after lock is related to the fact that in the same line a variable is declared. Personally I preffer more automatic construction that looks like this:

```
SPINLOCK_BLOCK(&lock) {
   do anything.
  break and return may be used here
}
```

We could implement it and it would be much nicer then.

Reply:
I am against the artificial `} while (0);` here. I meant only to keep things logically & visually consistent.
Since we have checkpatch limitation we could do.
```
key = lock();

logically_connected_stuff;
logically_connected_stuff;
logically_connected_stuff;
logically_connected_stuff;
logically_connected_stuff;

unlock(key);
```

What you are looking for is here https://github.com/nrfconnect/sdk-zephyr/blob/master/kernel/include/kernel_internal.h#L220

I don't know  why it is kept internal. Maybe we could (ab)use it... In the end we are part of the kernel anyway... at least we are in kernel space ;)

Reply:
Well.... in GCC it can really be done better and more safe  for the user: hint - use cleanup attribute of the variable. Then you may even call return from such a block and you are still on safe side.